python/taichi/lang/common_ops.py

```diff
@@ -1,61 +1,135 @@
+def numpy_or_constant(x):
+    import taichi as ti
+    if ti.is_taichi_class(x):
+        return x.to_numpy()
+    else:
+        return x
+
+
 class TaichiOperations:
     def __neg__(self):
+        if self.is_pyconstant():
```

Comment:
Does this still work inside the Taichi scope? `to_numpy()` actually launches a Taichi kernel https://github.com/taichi-dev/taichi/blob/43b1ccc68ded4d06dc383697be82fafea631c802/python/taichi/lang/matrix.py#L503

Reply:
In Taichi-scope, is_pyconstant will return False since it's a matrix of exprs.